# In this file the three dataset "us_derived_annual", "macroeconomics_variables" and "us-shareprice_daily" are merged into 1 dataset ("final dataset").

## All the datasets are opened and irrelevant features are deleted 

In [5]:
import numpy as np
import pandas as pd
import os
import missingno as msno

In [6]:
os.getcwd()
os.chdir("C:\\Users\\Jacco Bruin .LAPTOP-ACFSBHIL\\THESIS\\dataset simfiz")

In [7]:
df1 = pd.read_csv('us-derived-annual.csv', ";")
df2 = pd.read_csv('macro.csv')
df3 = pd.read_csv("us-shareprices-daily.csv", ";")
df3.head()

Ticker  SimFinId        Date    Open     Low    High   Close  Adj. Close  \
0   GOOG        18  2014-03-27  568.00  552.92  568.00  558.46      558.46   
1   GOOG        18  2014-03-28  561.20  558.67  566.43  559.99      559.99   
2   GOOG        18  2014-03-31  566.89  556.93  567.00  556.97      556.97   
3   GOOG        18  2014-04-01  558.71  558.71  568.45  567.16      567.16   
4   GOOG        18  2014-04-02  565.11  562.19  604.83  567.00      567.00   

   Dividend  Volume  Shares Outstanding  
0       NaN   13100         336050831.0  
1       NaN   41100         336050831.0  
2       NaN   10800         674462000.0  
3       NaN    7900         674462000.0  
4       NaN  146700         674462000.0

In [8]:
df1 = df1.drop(['Dividends Per Share','SimFinId','Currency', 'Fiscal Period', 'Report Date', 'Publish Date', 'Restated Date'],  axis = 1)
df2 = df2.drop(['Country', 'Code', 'ContinentCode'],  axis = 1)

In [9]:
df1_uni = np.unique(df1["Ticker"])
df1_uni

#y = ['2008',"2009","2010",'2011','2012','2013','2014','2015','2016','2017','2018','2019','2020']
y = [2008,2009,2010,2011,2012,2013,2014,2015,2016,2017,2018,2019,2020]
y = pd.DataFrame(y)
#print(y)

df1.index[df1['Ticker'] == "A"].tolist()
df3 = df1.iloc[0:13]
y = df3["Fiscal Year"]
y = y.iloc[0:13]
print(y)

0     2008
1     2009
2     2010
3     2011
4     2012
5     2013
6     2014
7     2015
8     2016
9     2017
10    2018
11    2019
12    2020
Name: Fiscal Year, dtype: int64


## Not all companies have the same amount of data. Some companies have data from 2008 till 2020, and some have data from 2015 till 2020. To select the most complete data only companies who had data from 2010 till 2020 were selected. 

In [10]:
#select only the year 2010 till 2020. 


df1_uni = np.unique(df1["Ticker"])

data = pd.DataFrame()

for i in df1_uni:
    x = df1[df1["Ticker"] == i]
    min_value = x["Fiscal Year"].min()
    #print(min_value)
    if min_value == 2008:    # i also want to select companies who have data from 2008 till 2020, therefore this line of code 
        min_value += 2
        x = x.drop(x[x["Fiscal Year"] == 2008].index)
    if min_value == 2009:    # i also want to select companies who have data from 2009 till 2020, therefore this line of code 
        min_value += 1
        x = x.drop(x[x["Fiscal Year"] == 2009].index)
    if (min_value == 2010) and (x["Fiscal Year"].max() == 2020) and (len(x["Fiscal Year"]) == 11):
        data = pd.concat([data, x])
     
    
print(len(data))      
pd.DataFrame(data)   

9097


Ticker  Fiscal Year        EBITDA    Total Debt  Free Cash Flow  \
21       AAL         2010  1.275000e+09  1.113600e+10   -3.780000e+08   
22       AAL         2011  4.900000e+07  8.220000e+09   -1.067000e+09   
23       AAL         2012  1.163000e+09  8.535000e+09   -1.039000e+09   
24       AAL         2013  2.419000e+09  1.679900e+10   -1.297000e+09   
25       AAL         2014  5.762000e+09  1.772000e+10   -5.650000e+08   
...      ...          ...           ...           ...             ...   
22449   ZYXI         2016  1.075000e+06  2.540000e+05             NaN   
22450   ZYXI         2017  9.230000e+06  3.540000e+05             NaN   
22451   ZYXI         2018  1.081800e+07  2.967000e+06             NaN   
22452   ZYXI         2019  1.184400e+07  0.000000e+00    8.949000e+06   
22453   ZYXI         2020  1.182100e+07  0.000000e+00   -3.040000e+05   

       Gross Profit Margin  Operating Margin  Net Profit Margin  \
21                 0.23031           0.01389           -0.02124   
22                 0.18237          -0.04325           -0.08253   
23                 0.20563           0.00595           -0.07548   
24                 0.25943           0.05231           -0.06858   
25                 0.28150           0.09962            0.06757   
...                    ...               ...                ...   
22449              0.73582           0.04807            0.00518   
22450              0.79434           0.38170            0.31431   
22451              0.81082           0.32491            0.29928   
22452              0.80617           0.24336            0.20874   
22453              0.78262           0.12792            0.11325   

       Return on Equity  Return on Assets  Free Cash Flow to Net Income  \
21              0.11939          -0.01877                       0.80255   
22              0.27830          -0.08298                       0.53916   
23              0.23488          -0.07980                       0.55384   
24              0.67155          -0.04338                       0.70720   
25              1.42603           0.06667                      -0.19604   
...                 ...               ...                           ...   
22449          -0.01815           0.01687                           NaN   
22450           1.50245           0.82484                           NaN   
22451           1.02809           0.49618                           NaN   
22452           0.48298           0.33568                       0.94279   
22453           0.15960           0.12572                      -0.03350   

       Current Ratio  Liabilities to Equity Ratio  Debt Ratio  \
21           0.78842                     -7.35944     0.44388   
22           0.78297                     -4.35368     0.34468   
23           0.78482                     -3.94353     0.36304   
24           1.03745                    -16.48078     0.39735   
25           0.87660                     20.38793     0.40995   
...              ...                          ...         ...   
22449        0.44183                     -2.07601     0.06209   
22450        2.07872                      0.82150     0.03965   
22451        2.05127                      1.07201     0.15412   
22452        4.34212                      0.43881     0.00000   
22453        6.22942                      0.26946     0.00000   

       Earnings Per Share, Basic  Earnings Per Share, Diluted  \
21                      -1.41441                     -1.41441   
22                     -15.83390                    -15.83390   
23                      -7.51934                     -7.51934   
24                      -6.54502                     -6.54502   
25                       4.01697                      3.92634   
...                          ...                          ...   
22449                    0.00221                      0.00221   
22450                    0.22904                      0.22186   
22451                    0.29388                      0.2

In [11]:
data_uni = np.unique(data["Ticker"])  
len(data_uni)                           #this gave 827 companies 

827

In [12]:
data.to_csv("dataset_2010_2020.csv", index = False)   # dataset is saved

### So, till here dataset of 827 companies (2010 till 2020) (length of 9097). Now add macro economische variable 

In [13]:
df6 = pd.read_csv('dataset_2010_2020.csv')
df2 = pd.read_csv('macro.csv')
df2 = df2.drop(['Country', 'Code', 'ContinentCode'],  axis = 1)
df2.head()
len(df2)

13

In [14]:
# we dont need the data 2008 till 2009, since companies only have data from 2010 till 2020. so I deleted these years.

df2 = df2.drop(df2[df2["Year"] == 2008].index)
df2 = df2.drop(df2[df2["Year"] == 2009].index)
df2

Year  Economic growth: the rate of change of real GDP  \
2   2010                                             2.56   
3   2011                                             1.55   
4   2012                                             2.25   
5   2013                                             1.84   
6   2014                                             2.53   
7   2015                                             3.08   
8   2016                                             1.71   
9   2017                                             2.33   
10  2018                                             3.00   
11  2019                                             2.16   
12  2020                                            -3.49   

    GDP per capita current U.S. dollars  \
2                              48466.66   
3                              49882.56   
4                              51602.93   
5                              53106.54   
6                              55049.99   
7                              56863.37   
8                              58021.40   
9                              60109.66   
10                             63064.42   
11                             65279.53   
12                             63543.58   

    Inflation: percent change in the Consumer Price Index  Unemployment rate  \
2                                                 1.6                   9.63   
3                                                 3.2                   8.95   
4                                                 2.1                   8.07   
5                                                 1.5                   7.38   
6                                                 1.6                   6.17   
7                                                 0.1                   5.28   
8                                                 1.3                   4.87   
9                                                 2.1                   4.36   
10                                                2.4                   3.90   
11                                                1.8                   3.67   
12                                                1.2                   8.31   

    Net portfolio equity inflows  Economic growth forecast  \
2                      178952.99                      2.71   
3                      123357.00                      1.55   
4                      239066.01                      2.28   
5                      -62642.00                      1.84   
6                      154312.00                      2.29   
7                     -187306.00                      2.71   
8                     -139700.99                      1.67   
9                      149634.01                      2.26   
10                     156916.01                      2.92   
11                    -244072.01                      2.29   
12                     725723.97                     -3.41   

    Inflation forecast  Unemployment rate forecast  \
2                 1.69                        9.61   
3                 3.09                        8.93   
4                 1.82                        8.08   
5                 1.32                        7.36   
6                 0.52                        6.16   
7                 0.70                        5.28   
8                 2.18                        4.88   
9                 2.17                        4.35   
10                1.91                        3.89   
11                2.06                        3.68   
12                1.55                        8.11   

    Political stability index (-2.5 weak; 2.5 strong)  \
2                                                0.44   
3                                                0.59   
4                                                0.63   
5                                                0.64   
6                                                0.58   
7                                                0.68   


In [15]:
#WATCH OUT, ONLY RUN THIS ONCE, OTHER WISE SUPER BIG DATASET
# 
df2.reset_index(drop=True, inplace=True)    #reset index otherwise appending is not working properly
df2 = df2.append([df2]*826,ignore_index=True) # multiply the macroeconomic dataset by amount of companies, so datasets are equal 

In [16]:
frames = [df6, df2]
new_data2 = pd.concat(frames, axis = 1)    #merge the financial ratio dataset with macroeconomic variables

In [17]:
new_data2 = new_data2.drop(columns = "Year")  
new_data2 = new_data2.rename(columns = {"Fiscal Year": "year"})  # year was two times in the dataset, so one deleted, and one I give a new name

In [19]:
new_data2.to_csv("_macro_dataset_2010_2020.csv", index = False)

### So far dataset from year 2010 to 2020 with macro economic variables. In the next block im gonna add the stock prices 

In [63]:
df7 = pd.read_csv("_macro_dataset_2010_2020.csv")
df3 = pd.read_csv("us-shareprices-daily.csv", ";")  # this is a really big dataset, so be carefull when you run this

     Ticker  year        EBITDA    Total Debt  Free Cash Flow  \
0       AAL  2010  1.275000e+09  1.113600e+10   -3.780000e+08   
1       AAL  2011  4.900000e+07  8.220000e+09   -1.067000e+09   
2       AAL  2012  1.163000e+09  8.535000e+09   -1.039000e+09   
3       AAL  2013  2.419000e+09  1.679900e+10   -1.297000e+09   
4       AAL  2014  5.762000e+09  1.772000e+10   -5.650000e+08   
...     ...   ...           ...           ...             ...   
9092   ZYXI  2016  1.075000e+06  2.540000e+05             NaN   
9093   ZYXI  2017  9.230000e+06  3.540000e+05             NaN   
9094   ZYXI  2018  1.081800e+07  2.967000e+06             NaN   
9095   ZYXI  2019  1.184400e+07  0.000000e+00    8.949000e+06   
9096   ZYXI  2020  1.182100e+07  0.000000e+00   -3.040000e+05   

      Gross Profit Margin  Operating Margin  Net Profit Margin  \
0                 0.23031           0.01389           -0.02124   
1                 0.18237          -0.04325           -0.08253   
2                 0.2

In [21]:
df3['year'] = pd.DatetimeIndex(df3['Date']).year      #change date to year
df3['month'] = pd.DatetimeIndex(df3['Date']).month    #change data to month

In [22]:
df3_uni = np.unique(df3["Ticker"])
len(df3_uni)  #is contains 2647 companies

2647

# this code get oktober month for every year. takes long to run (around 20 min). therefore in markdown

data = pd.DataFrame()

sum = 1

for i in df3_uni:
    x = df3[df3["Ticker"] == i]
    month = x[x["month"] == 10]
    year = month.drop_duplicates('year')
    print(sum)
    sum += 1
    data = pd.concat([data, year])
     
    
data.to_csv("dataset_only_october.csv", index = False)

## I now have only selected the month october 

In [36]:
df8 = pd.read_csv("dataset_only_october.csv")
df8_1 = df8[["Ticker","Open", "year"]]

In [68]:
# append the openingsprice to the financial ratio + macroeconomic variable model. So only years 2010 till 2020 are selected 

df7_uni = np.unique(df7["Ticker"])

data = pd.DataFrame()

for i in df7_uni:
    x = df7[df7["Ticker"] == i]
    y = df8_1[df8_1["Ticker"] == i]
    total_data = x.merge(y, on = "year", how = "left")
    data = pd.concat([data, total_data])

data.to_csv("tete.csv")  # save the dataset, so result till now are saved

In [69]:
df10 = pd.read_csv('tete.csv')  #open the dataset again

In [77]:
# add the percentage difference between 2010 - 2011, 2010 - 2012, 2010 = 2013 etc

df10_uni = np.unique(df10["Ticker_x"])
data = pd.DataFrame()

for i in df10_uni:
    
    x = df10[df10["Ticker_x"] == i]
    price_2010 = x["Open"].iloc[0]
    total_result = []
    for i in x["Open"]:
        result = round((i - price_2010) / price_2010,4)
        total_result.append(result)
    x["profit/loss"] =  total_result
    data = pd.concat([data, x])

data.to_csv("TEST_7_11_21.csv")    # save the dataset, so result till now are saved

C:\Users\Jacco Bruin .LAPTOP-ACFSBHIL\anaconda3\envs\tens\lib\site-packages\ipykernel_launcher.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [78]:
df11 = pd.read_csv('TEST_7_11_21.csv')   #open the dataset again

Unnamed: 0  Unnamed: 0.1 Ticker_x  year        EBITDA    Total Debt  \
0              0             0      AAL  2010  1.275000e+09  1.113600e+10   
1              1             1      AAL  2011  4.900000e+07  8.220000e+09   
2              2             2      AAL  2012  1.163000e+09  8.535000e+09   
3              3             3      AAL  2013  2.419000e+09  1.679900e+10   
4              4             4      AAL  2014  5.762000e+09  1.772000e+10   
...          ...           ...      ...   ...           ...           ...   
9092        9092             6     ZYXI  2016  1.075000e+06  2.540000e+05   
9093        9093             7     ZYXI  2017  9.230000e+06  3.540000e+05   
9094        9094             8     ZYXI  2018  1.081800e+07  2.967000e+06   
9095        9095             9     ZYXI  2019  1.184400e+07  0.000000e+00   
9096        9096            10     ZYXI  2020  1.182100e+07  0.000000e+00   

      Free Cash Flow  Gross Profit Margin  Operating Margin  \
0      -3.780000e+08              0.23031           0.01389   
1      -1.067000e+09              0.18237          -0.04325   
2      -1.039000e+09              0.20563           0.00595   
3      -1.297000e+09              0.25943           0.05231   
4      -5.650000e+08              0.28150           0.09962   
...              ...                  ...               ...   
9092             NaN              0.73582           0.04807   
9093             NaN              0.79434           0.38170   
9094             NaN              0.81082           0.32491   
9095    8.949000e+06              0.80617           0.24336   
9096   -3.040000e+05              0.78262           0.12792   

      Net Profit Margin  ...  Unemployment rate  Net portfolio equity inflows  \
0              -0.02124  ...               9.63                     178952.99   
1              -0.08253  ...               8.95                     123357.00   
2              -0.07548  ...               8.07                     239066.01   
3              -0.06858  ...               7.38                     -62642.00   
4               0.06757  ...               6.17                     154312.00   
...                 ...  ...                ...                           ...   
9092            0.00518  ...               4.87                    -139700.99   
9093            0.31431  ...               4.36                     149634.01   
9094            0.29928  ...               3.90                     156916.01   
9095            0.20874  ...               3.67                    -244072.01   
9096            0.11325  ...               8.31                     725723.97   

      Economic growth forecast  Inflation forecast  \
0                         2.71                1.69   
1                         1.55                3.09   
2                         2.28                1.82   
3                         1.84                1.32   
4                         2.29                0.52   
...                        ...                 ...   
9092                      1.67                2.18   
9093                      2.26                2.17   
9094                      2.92                1.91   
9095                      2.29                2.06   
9096                     -3.41                1.55   

      Unemployment rate forecast  \
0                           9.61   
1                           8.93   
2                           8.08   
3                           7.36   
4                           6.16   
...                          ...   
9092                        4.88   
9093                        4.35   
9094                        3.89   
9095                        3.68   
9096                        8.11   

      Political stability index (-2.5 weak; 2.5 strong)  \
0                                                  0.44   
1                                                  0.59   
2                                                  0.63   
3                                

In [80]:
df11 = df11.iloc[: , 1:]  # run this three time to keep deleting the first columns, which are irrelevant index columns

Ticker_x  year        EBITDA    Total Debt  Free Cash Flow  \
0         AAL  2010  1.275000e+09  1.113600e+10   -3.780000e+08   
1         AAL  2011  4.900000e+07  8.220000e+09   -1.067000e+09   
2         AAL  2012  1.163000e+09  8.535000e+09   -1.039000e+09   
3         AAL  2013  2.419000e+09  1.679900e+10   -1.297000e+09   
4         AAL  2014  5.762000e+09  1.772000e+10   -5.650000e+08   
...       ...   ...           ...           ...             ...   
9092     ZYXI  2016  1.075000e+06  2.540000e+05             NaN   
9093     ZYXI  2017  9.230000e+06  3.540000e+05             NaN   
9094     ZYXI  2018  1.081800e+07  2.967000e+06             NaN   
9095     ZYXI  2019  1.184400e+07  0.000000e+00    8.949000e+06   
9096     ZYXI  2020  1.182100e+07  0.000000e+00   -3.040000e+05   

      Gross Profit Margin  Operating Margin  Net Profit Margin  \
0                 0.23031           0.01389           -0.02124   
1                 0.18237          -0.04325           -0.08253   
2                 0.20563           0.00595           -0.07548   
3                 0.25943           0.05231           -0.06858   
4                 0.28150           0.09962            0.06757   
...                   ...               ...                ...   
9092              0.73582           0.04807            0.00518   
9093              0.79434           0.38170            0.31431   
9094              0.81082           0.32491            0.29928   
9095              0.80617           0.24336            0.20874   
9096              0.78262           0.12792            0.11325   

      Return on Equity  Return on Assets  ...  Unemployment rate  \
0              0.11939          -0.01877  ...               9.63   
1              0.27830          -0.08298  ...               8.95   
2              0.23488          -0.07980  ...               8.07   
3              0.67155          -0.04338  ...               7.38   
4              1.42603           0.06667  ...               6.17   
...                ...               ...  ...                ...   
9092          -0.01815           0.01687  ...               4.87   
9093           1.50245           0.82484  ...               4.36   
9094           1.02809           0.49618  ...               3.90   
9095           0.48298           0.33568  ...               3.67   
9096           0.15960           0.12572  ...               8.31   

      Net portfolio equity inflows  Economic growth forecast  \
0                        178952.99                      2.71   
1                        123357.00                      1.55   
2                        239066.01                      2.28   
3                        -62642.00                      1.84   
4                        154312.00                      2.29   
...                            ...                       ...   
9092                    -139700.99                      1.67   
9093                     149634.01                      2.26   
9094                     156916.01                      2.92   
9095                    -244072.01                      2.29   
9096                     725723.97                     -3.41   

      Inflation forecast  Unemployment rate forecast  \
0                   1.69                        9.61   
1                   3.09                        8.93   
2                   1.82                        8.08   
3                   1.32                        7.36   
4                   0.52                        6.16   
...                  ...                         ...   
9092                2.18                        4.88   
9093                2.17                        4.35   
9094                1.91                        3.89   
9095                2.06                        3.68   
9096                1.55                        8.11   

      Political stability index (-2.5 weak; 2.5 strong)  \
0                                                  0.44   
1                             

In [81]:
df11.drop(columns=df11.columns[30],    # ticker is now twice in the dataset. So the last one is deleted here
        axis=1, 
        inplace=True)

In [83]:
df11.to_csv("dataset_8_11_21.csv")  # the final dataset for this file

### I have now the dataset with everything I need. In the file "deal_with_missing_values_part_1" I will deal with missing values. 